addon/mixins/adapter-fetch.ts

```diff
@@ -0,0 +1,187 @@
+import Mixin from '@ember/object/mixin';
+import { assign } from '@ember/polyfills';
+import RSVP, { reject } from 'rsvp';
+import fetch from 'fetch';
+import mungOptionsForFetch from '../utils/mung-options-for-fetch';
+import determineBodyPromise from '../utils/determine-body-promise';
+import DS from 'ember-data';
+import Mix from '@ember/polyfills/types';
+import { get } from '@ember/object';
+import {
+  PlainObject,
+  PlainHeaders,
+  Method,
+  FetchOptions,
+  Nullable,
+  AjaxOptions
+} from 'ember-fetch/types';
+
+/**
+ * Helper function to create a plain object from the response's Headers.
+ * Consumed by the adapter's `handleResponse`.
+ */
+export function headersToObject(headers: Headers): PlainObject {
+  let headersObject: PlainObject = {};
+
+  if (headers) {
+    headers.forEach((value, key) => (headersObject[key] = value));
+  }
+
+  return headersObject;
+}
+
+export default Mixin.create({
```

Comment:
I wanted to do `Mixin.create<TheMixinTypeLiteral, DS.RestAdapter>` here. But it seems it's not correct. For example `this.ajaxOptions` might be wrapped by base adapter's `this.ajaxOptions`'s `this._super(...arguments)` and return not necessarily `FetchOptions`.

Reply:
from a typescript standpoint, there's no "clean" way to declare assumptions about the type of object the mixin will be applied to (this is one of the general pitfalls of using mixins in general).

The best you could do is explicitly specify `this` in your method signatures. For example

```ts
ajaxOptions(this: DataAdapter, url: string, type: Method, options: object): FetchOptions {
  // in here `this` will be a DataAdapter
  ...
} 

```